analyzers/src/SonarAnalyzer.CSharp/Helpers/CSharpEquivalenceChecker.cs

```diff
@@ -36,6 +36,33 @@ internal static class CSharpEquivalenceChecker
         public static bool AreEquivalent(SyntaxList<SyntaxNode> nodeList1, SyntaxList<SyntaxNode> nodeList2) =>
             Common.EquivalenceChecker.AreEquivalent(nodeList1, nodeList2, NodeComparator);
 
+        public static bool AreEquivalentWithEqualInvocations(SyntaxList<SyntaxNode> nodeList1, SyntaxList<SyntaxNode> nodeList2, SemanticModel model)
+        {
+            if (nodeList1.Count != nodeList2.Count)
+            {
+                return false;
+            }
+
+            for (var i = 0; i < nodeList1.Count; i++)
+            {
+                if (!AreEquivalent(nodeList1[i], nodeList2[i]))
+                {
+                    return false;
+                }
+                if (FindInvocationExpression(nodeList1[i]) is { } invocation1
+                    && FindInvocationExpression(nodeList2[i]) is { } invocation2
+                    && !invocation1.IsEqualTo(invocation2, model))
```

Comment:
this has the potential to get very expensive. What do you think about checking the full syntax and only when the syntax is similar retrieve all the invocations and check them at the symbol level one by one?

Reply:
I would approach this a bit differently. In the rule call 2 methods: `AreEquivalent` as before and only if that returns true, then call this method that will check only the invocation symbols.

Something similar with: 
```
public static bool HaveTheSameInvocations(SyntaxList<SyntaxNode> first, SyntaxList<SyntaxNode> second, SemanticModel model)
        {
            var referenceInvocations = first.SelectMany(x => x.DescendantNodes().OfType<InvocationExpressionSyntax>()).ToArray();
            var candidateInvocations = second.SelectMany(x => x.DescendantNodes().OfType<InvocationExpressionSyntax>()).ToArray();
            if (referenceInvocations.Length != candidateInvocations.Length)
            {
                return false;
            }

            for (var i = 0; i < referenceInvocations.Length; i++)
            {
                if (!referenceInvocations[i].IsEqualTo(candidateInvocations[i], model))
                {
                    return false;
                }
            }
            return true;
        }
```

Two main advantages:
- we check the full syntax without retrieving any symbols. This part is cheap.
- we avoid code duplication and we check the invocation symbols separately.
- we check all the invocations. in the previous implementation, only the first one was considered.

Reply:
This seems much better indeed - code used only in this rule, is placed in this rule.